### PR TITLE
Resolve 10342 : [Ltac2] Add array library

### LINKED ISSUE #10342: [Ltac2] Add array library (compatible with OCaml standard library)
Ltac2 has not yet an array library. Since arrays are used a lot in type "kind", this is a bit inconvenient. I am about to prepare a PR to add one.

I decided to make it as compatible as possible to the OCaml standard library to easy conversion of Ltac2 programs to OCaml programs. Besides different low level functions, the only difference is that mem takes an equality function, since Ltac2 doesn't seem to have a generic equality function.

### LINKED ISSUE #10348: [Ltac2] no way to create an empty array without supplying an instance
Ltac2 has no way to create empty arrays without supplying an instance. This leads to a lot of additional checks in the array library I proposed in PR #10343, since for functions which should return an empty array one frequently has no instance available one could supply to make.

### PR DESCRIPTION
**Pull request by MSoegtropIMC**
This PR adds an array library for Ltac2.

I copied the Invalid_argument exception from PR #10219, so that it passes CI. I will remove it if 10219 is merged before.

The `mem` function is not compatible with the OCaml standard library, it requires an equality function since Ltac2 doesn't seem to have a generic equality functions. Otherwise it should be compatible except for names of low level functions.

I generally made the recursive low level functions global and generic in the sense that they can also be applied to parts of an array, in case someone needs this. In some cases this might lead to minor performance losses for handling both a current index and a length.

I have test cases for all high level function, but I am not sure where I would put these.

Fixes #10342.
Fixes #10348.

### CODE REVIEW

File: user-contrib/Ltac2/Array.v
Comment (gares) [inline user-contrib/Ltac2/Array.v]: @ppedrot is `set` unsafe? (as in no boundary check?)

Comment (ppedrot) [reply user-contrib/Ltac2/Array.v]: Nope. It'll throw an out-of-bounds panic if misused.

Comment (gares) [reply user-contrib/Ltac2/Array.v]: The these functions shall not be called unsafe I guess

File: user-contrib/Ltac2/Array.v
Comment (ppedrot) [inline user-contrib/Ltac2/Array.v]: You should define a helper function `invalid_argument : string -> 'a` for this sequence of compositions, it seems to be used quite a bit in the rest. It should probably live in `Init`.

File: user-contrib/Ltac2/Array.v
Comment (ppedrot) [inline user-contrib/Ltac2/Array.v]: I don't think that this function makes sense. Already in OCaml it is quite dubious.

File: user-contrib/Ltac2/Array.v
Comment (ppedrot) [inline user-contrib/Ltac2/Array.v]: So we're missing the `empty` function. It should be implemented before merging this PR, otherwise we'll get fancy behaviours as witnessed below.

File: user-contrib/Ltac2/Array.v
Comment (ppedrot) [inline user-contrib/Ltac2/Array.v]: IIRC the OCaml stdlib has a fast, C implementation for this. I guess we can just reflect the OCaml function in Ltac2, I'll do it with the addition of `empty`.

**Comment by ppedrot** [comment]
> The mem function is not compatible with the OCaml standard library, it requires an equality function since Ltac2 doesn't seem to have a generic equality functions.

This is actually a defect of the OCaml API, at least according to me. We're struggling not to rely on the generic equality in OCaml, because it's a real can of worms. It breaks opacity invariants and can behave nonsensically sometimes (most notably for mutable data).

File: user-contrib/Ltac2/Array.v
Comment (MSoegtropIMC) [reply user-contrib/Ltac2/Array.v]: I though about having a function which includes the match on bool, so something like `assert_invalid_argument : bool -> string -> unit` which either throws when the first argument is false or does nothing when it is true. This would allow to have more specific tests and error messages without writing 20 lines of code. Some functions have longish orb chains which I would split in several tests with specific messages.

As is the error messages follow closely what OCaml does, but since debugging in OCaml might be nicer, it would make sense to have nicer error messages in Ltac2.

File: user-contrib/Ltac2/Array.v
Comment (MSoegtropIMC) [reply user-contrib/Ltac2/Array.v]: Removing it would be the only deviation from the OCaml library. I think it doesn't hurt to keep it and I might even have uses for it in automation for score tables.

File: user-contrib/Ltac2/Array.v
Comment (MSoegtropIMC) [reply user-contrib/Ltac2/Array.v]: I am looking forward to it :-)

File: user-contrib/Ltac2/Array.v
Comment (MSoegtropIMC) [reply user-contrib/Ltac2/Array.v]: It is a good question how checks should be handled - possibly I was following to much the OCaml library. The idea is that the "safe" functions give nicer error message at least mentioning the name of the function which failed. For the low level "unsafe" functions give less specific errors. The options are:

- leave all checks away and call the recursive helper functions _aux or _lowlevel. Performance wise this would be best and I don't know (yet) how helpful the additional error messages are.
- keep it as is with more specific error messages, possibly renaming the low level functions as suggested above.

**Comment by MSoegtropIMC** [comment]
P.S.: CI fails because also the Bool library from #10219 is missing (I am using orb). I will wait until this is merged before I take care of CI, unless this would take a long time.

**Comment by MSoegtropIMC** [comment]
I fixed the issues mentioned above. Most notably I added an empty, blit and fill function to Ltac2. This are all functions which iterate over the array and don't require an Ltac2 function as argument. The Ltac2 code has been adjusted to allow empty arrays and to use the new low level functions.

Note: CI still won't pass because I use the Bool and Int extensions from Jason's PR.

**Comment by ppedrot** [comment]
@MSoegtropIMC can you rebase on a fresh master?

**Comment by MSoegtropIMC** [comment]
@ejgallego : can you help me with the Azure Windows failure? It just says:
```
Cmd.exe exited with code '1'.
```
I am a bit lost ...

**Comment by SkySkimmer** [comment]
> It just says:

at https://dev.azure.com/coq/coq/_build/results?buildId=3845&view=logs&jobId=2d2b3007-3c5c-5840-9bb0-2b1ea49925f3 I see
~~~
Exception calling "DownloadFile" with "2" argument(s): "Unable to connect to the remote server"
At line:1 char:1
+ (New-Object Net.WebClient).DownloadFile('http://www.cygwin.com/setup- ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
    + FullyQualifiedErrorId : WebException
 
'setup-x86_64.exe' is not recognized as an internal or external command,
operable program or batch file.
The system cannot find the path specified.
##[error]Cmd.exe exited with code '1'.
~~~
ie the cygwing download failed (probably a transient error)

**Comment by MSoegtropIMC** [comment]
@SkySkimmer : thanks! I missed the tiny grayish "more information" link.

**Comment by MSoegtropIMC** [comment]
@ppedrot : please review this already while I take care of the minor CI issues.

File: user-contrib/Ltac2/Array.v
Comment (ppedrot) [inline user-contrib/Ltac2/Array.v]: The API doesn't let you do that (yet) but this primitive shouldn't be a closure, it should return `'a array` directly as this is not a problematic value w.r.t. sharing. I'll write a preliminary patch to let you write this.

Comment (MSoegtropIMC) [reply user-contrib/Ltac2/Array.v]: > this primitive shouldn't be a closure, it should return `'a array` directly as this is not a problematic value w.r.t. sharing

I am not deep enough in OCaml internals to know if the "CA" in OCaml goes so far that arrays of different types can share the same binary object. Interesting to know that this is the case.

One more point: I am not sure if it is required to make empty a primitive. One should be able to construct an empty function from Array.init and a throw. Of cause one still might want to have it but I wonder why the OCaml library doesn't have an Array.empty function.

Would it be agreeable to add the string functions from my system function PR to this PR, that is:<https://github.com/coq/coq/pull/10370/commits/d67fce58ea66fffc47cd755f50a410ea7ce3be46>. I also need this at other places without system functions and I guess others might need this as well. I would add conversion in both directions then.

**Comment by SkySkimmer** [comment]
ocaml has `[||]` to produce empty arrays

Gaëtan Gilbert

File: user-contrib/Ltac2/Array.v
Comment (ppedrot) [reply user-contrib/Ltac2/Array.v]: > One more point: I am not sure if it is required to make empty a primitive. One should be able to construct an empty function from Array.init and a throw.

No, you can't do that. If you throw inside init, you'll get an exception due to the call-by-value semantics.

**Comment by MSoegtropIMC** [comment]
> ocaml has `[||]` to produce empty arrays

Good point - I was just looking at <https://caml.inria.fr/pub/docs/manual-ocaml/libref/Array.html> and forgot the built-in functionality. Can you think of other primitive/built-in array functionality in OCaml I missed?

File: user-contrib/Ltac2/Array.v
Comment (MSoegtropIMC) [reply user-contrib/Ltac2/Array.v]: > No, you can't do that. If you throw inside init, you'll get an exception due to the call-by-value semantics.

I meant a thunked throw, so it it never gets called for an empty array. Actually this does work:
```
Require Import Ltac2.Init.
Require Import Ltac2.Message.
Require Import Ltac2.Array.

Ltac2 print2 m1 m2 := print (Message.concat m1 m2).
Ltac2 print3 m1 m2 m3 := print2 m1 (Message.concat m2 m3).

Ltac2 print_int_array a :=
  iteri (fun i x => print3 (of_int i) (of_string "=") (of_int x)) a.

Ltac2 myempty _ := init 0 (fun _ => (Control.throw_invalid_argument "myempty: internal")).

Goal True.
  let a := myempty () in print_int_array a.
  let a := append (myempty ()) (init 4 (fun i => (Int.add i 10))) in print_int_array a.
```

Comment (SkySkimmer) [reply user-contrib/Ltac2/Array.v]: Look at the definition of init, it uses empty inside so all you did was define `myempty := empty`

Comment (MSoegtropIMC) [reply user-contrib/Ltac2/Array.v]: Ok, that was indeed silly - I forgot that `init` is not a primitive function.

Comment (MSoegtropIMC) [reply user-contrib/Ltac2/Array.v]: @silene : how shall we proceed here? For me as a user it is not so bad that empty is a function while it could be a value if Ltac2 would support that. A lack of standard array functions in general is much more of an issue for me. Maybe we can for the time being rename it to say `lowlevel_empty` to indicate that it is not intended to be used directly, or that people have to expect changes? If it has a name like this it would be no issue to search and replace `lowlevel_empty ()` with empty later.

File: user-contrib/Ltac2/Array.v
Comment (MSoegtropIMC) [reply user-contrib/Ltac2/Array.v]: I added already in #10219 a functions like Control.assert_valid_argument and use them throughout to give fine grain error messages.

File: user-contrib/Ltac2/Array.v
Comment (MSoegtropIMC) [reply user-contrib/Ltac2/Array.v]: I added the empty array function and fixed all the 0 cases.

File: user-contrib/Ltac2/Array.v
Comment (MSoegtropIMC) [reply user-contrib/Ltac2/Array.v]: And indeed it is much simpler this way. I first though that I need some translation from Coq lists to OCaml lists, but apparently this is already built into Ltac2, so just these 3 lines:
```
let () = define1 "array_concat" (list block) begin fun l ->
  wrap (fun () -> v_blk 0 (Array.concat (List.map snd l)))
end
```
with the declaration
```
Ltac2 @external concat : ('a array) list -> 'a array := "ltac2" "array_concat".
```
do the trick - I like it ;-)

**Comment by MSoegtropIMC** [comment]
@ppedrot : I made concat an OCaml function as you suggested and it works as expected. But I have to leave it to you to fix the empty function - as you said this requires some extension to the framework. Right now empty always produces an int array (not sure why int and not anythign else or an error). I hope you find some time for this after 8.11.

File: user-contrib/Ltac2/Array.v
Comment (MSoegtropIMC) [reply user-contrib/Ltac2/Array.v]: > The API doesn't let you do that (yet) but this primitive shouldn't be a closure, it should return `'a array` directly as this is not a problematic value w.r.t. sharing. I'll write a preliminary patch to let you write this.

Ehmm sorry - I made a mistake in testing the empty function. Actually the empty function does what I would expect. See e.g. the following snippet:
```
Require Import Ltac2.Ltac2.
Import Ltac2.Message.
Import Ltac2.Array.

Ltac2 print2 m1 m2 := print (Message.concat m1 m2).
Ltac2 print3 m1 m2 m3 := print2 m1 (Message.concat m2 m3).

Ltac2 print_int_array a :=
  iteri (fun i x => print3 (of_int i) (of_string "=") (of_int x)) a.

Ltac2 of_bool b := match b with true=>of_string "true" | false=>of_string "false" end.

Ltac2 print_bool_array a :=
  iteri (fun i x => print3 (of_int i) (of_string "=") (of_bool x)) a.

Goal True.
  (* test empty with int *)
  let a := append (empty ()) (init 3 (fun i => (Int.add i 10))) in print_int_array a.
  let a := append (init 3 (fun i => (Int.add i 10))) (empty ()) in print_int_array a.

  (* test empty with bool *)
  let a := append (empty ()) (init 3 (fun i => (Int.ge i 2))) in print_bool_array a.
  let a := append (init 3 (fun i => (Int.ge i 2))) (empty ()) in print_bool_array a.

  (* test init with int *)
  let a := init 0 (fun i => (Int.add i 10)) in print_int_array a.
  let a := init 4 (fun i => (Int.add i 10)) in print_int_array a.

  (* test init with bool *)
  let a := init 0 (fun i => (Int.ge i 2)) in print_bool_array a.
  let a := init 4 (fun i => (Int.ge i 2)) in print_bool_array a.
```
So what does not work with the empty function - despite the fact that it shouldn't be a function but a value?

**Comment by ppedrot** [comment]
@MSoegtropIMC can you rebase this on a recent master? In the end, I think this can be merged as-is and refined by further PRs.

**Review by JasonGross — DISMISSED** [review]
I don't think making `user-contrib/Ltac2/tac2core.ml` executable (100644 → 100755) is correct...

**Comment by MSoegtropIMC** [comment]
> I don't think making `user-contrib/Ltac2/tac2core.ml` executable (100644 → 100755) is correct...

Thanks! I fixed this and Array.v as well.

**Comment by MSoegtropIMC** [comment]
@ppedrot : I also have a file with testcases. Should I add this to the commit and the Ltac make test or do you think this is not required?

**Comment by ppedrot** [comment]
Tests are always better to ship with the corresponding feature.

**Comment by MSoegtropIMC** [comment]
> Tests are always better to ship with the corresponding feature.

OK, I will add the test file tomorrow. Please wait with the merge.

**Comment by MSoegtropIMC** [comment]
@ppedrot : I add test test-file, so ready t go.

You might want to leave a note in the user-contrib/ltac2 folder on where examples and test cases are. It is a bit odd that this is under user-contrib and the tests and examples are under test-suite/ltac2. @SkySkimmer explained to me on gitter why this is, but it is not obvious to users looking for the esamples.

**Comment by MSoegtropIMC** [comment]
@ppedrot : the fiat crypto failure seems unrelated (it fails also for other PRs and some master commits, but not all). I fixed the filemode issues spotted by @JasonGross and also added a test file. So from my point of view this is good to go.

**Review by ppedrot — APPROVED** [review]

**Review by ppedrot — CHANGES_REQUESTED** [review]
@MSoegtropIMC sorry, I just realized that this PR is lacking a changelog. Can you add one please?

**Comment by MSoegtropIMC** [comment]
> I just realized that this PR is lacking a changelog. Can you add one please?

I wonder into which folder under doc/chaneglog this should go. I was looking at #10219 but this doesn't have one either - should I create a change log for that as well?

**Comment by ppedrot** [comment]
This should go in `05-tactic-language/`. Although the name is not pluralized, this doesn't read *Ltac1*.

**Review by ppedrot — APPROVED** [review]

**Pull request merged**
